### ISSUE TITLE
RFC: Variable size task outputs

### ISSUE THREAD

**Issue by magnatelee** [StanfordLegion/legion#880]
I have been dealing with tasks with variable size outputs for the past few months and I feel now it's a good time to draw everyone's attention to the issue again and discuss possible options. Here I describe three approaches that I know of, two of which require some API extensions. Hopefully, this will fire up some discussions and lead us to a reasonable solution.

---

#### Requirements for variable size outputs
 
Before I talk about the approaches, I'd like to summarize the requirements to handle variable size task outputs:

1. We should be able to allocate instances during task execution.
2. We should be able to return the allocated instances from the task in some way.
3. We should be able to incorporate the output instances back to Legion's data model whenever necessary (e.g., when needing to repartition the data).

I intentionally use "instances" instead of regions here, as regions are not necessarily the only way to communicate names of instances, although that is the case in the current API. You will soon see an example where instances are created and used without regions.

#### Two-pass scheme

The simplest way to deal with variable size outputs is to employ a two-pass scheme; we launch a "dry-run" task that simply computes the size of the output, and then we use the result to create the output region for the actual computation task. Thanks to the fix to #707, we no longer need to block on the dry-run task to create an index space of the output, although mapping of the computation task still needs to wait for the dry-run task to finish.

The biggest problem with this approach is that the dry-run tasks often need to duplicate computation. In many cases I've seen, I had to perform most, if not all, of the computation just to know the size a priori. Writing an efficient dry-run task is not always possible either, since we often need to use external libraries that expose entires only to the whole computation. For this reason, I don't think the two-pass scheme can be our story in the long term.

#### BLIS approach

Another approach we can take as of today is a BLIS-like approach. Here is how it goes: 

We first create an output region `R` of an unbounded size, say [0, INT64_MAX). Once the size is determined during task execution, we create a subregion `S` of `R` and inline map it to an instance (Requirement 1). The name of `S` can be either returned from the task in a future or referred to by some compile time constant (Requirement 2). Everything stays in Legion's data model with regions and subregions (Requirement 3) and no extra mechanism is necessary to implement this approach. In a pseudo code, the approach looks like this:

```
task producer(R) writes(R)
  ...
  S = create_subregion(R, [0, actual_size) )
  I = inline_map(S)
  -- resume computation with I
end

task main()
  R = region([0, INT64_MAX), some_type_T)
  producer(R)
  S = get_subregion(R, name_of_S)   ---- (1)
  consumer(S)
end
```

The major issue with the BLIS approach is that it introduces an undesirable dependency bewteen task execution and logical analysis. In the example above, there is a race between `producer` creating a subregion under the `name_of_S` and `main` retrieving the subregion with that name, and therefore they must be synchronized somehow. However, regardless of our choice of synchronization, the logical analysis of `consumer` will be blocked by `producer`'s execution, either directly or indirectly:

- If `consumer` is an index task, we need to make the projection functor to block on the future of `producer` before it walks down the region tree to get `S`, which happens in the logical analysis.
- If `consumer` is an individual task, there is no way other than blocking on `producer`'s future before line (1). This in turn blocks `consumer`'s logical analysis.
- We may consider introducing the old index requirement back, but still the logical analysis of any task that needs to traverse the index space tree of `R`'s index space must wait until `producer` is done mutating the tree.

Given the sequential nature of logical analysis, this kind of chaining can be quite detrimental to performance, especially if there are many tasks with variable size outputs.

A possible workaround to this issue is to create the `name_of_S` before launching `producer` so that `consumer` is unblocked in the logical analysis and instead blocked during mapping. In fact, we already have most of the necessary mechanisms to do this and the only missing piece is decoupling of names and values for futures. This example demonstrates the required API change:
```
task producer(R, size) writes(R), writes(size)                -- (1)
  ...
  size.set_result(Domain( [0, actual_size) ))                 -- (2)
  S = get_subregion(R, name_of_S)
  I = inline_map(S)                                           -- (3)              
  ...
  
task main()
  ...
  R = region([0, INT64_MAX], some_type_T)
  size = create_deferred_future()                             -- (4)
  P = create_pending_partition(R.index_space)
  create_index_space_intersection(P, name_of_S, ispace(size)) -- (5)
  
  producer(R, size)
  S = get_subregion(R)
  consumer(S)
end
```
The idea is to defer binding of a value to a future. A deferred future is created without any initial value (line (4)) and its ready event is not yet triggered. Otherwise, the deferred future can be used just like a normal future. For example, a deferred future `size` is used to create an index space at line (5). The value of the deferred future can be set with `set_result` in a task, which will also trigger the ready event (line (2)). The task must take the deferred future in write mode (line (1)) to be able to call `set_result` on it.

Both passing a deferred future in read mode and passing a normal future in write mode are illegal, and this kind of illegal uses can be easily detected dynamically. However, it doesn't seem possible to detect indirect out-of-order accesses to deferred futures; in the example above, nothing really prevents us from launching `consumer` first and then `producer` and hitting a deadlock. We may require the caller to always pass the future if the callee uses any index space created from that future, but this is somewhat hacky. So far I haven't found a good solution to this.

One more thing to think about is whether we should provide any APIs to better support the BLIS structure. Most notably, partitioning is one of the most common things we do in Legion, but since the BLIS structure fragments the data into non-contigous subspaces, independent partitions like equal partitions or weighted partitions are non-trivial to construct. In fact, it's often easier to first convert the BLIS into a contiguous region and then create a partition. Here is a code that I had to write to convert a BLIS-partitioned region into a contiguous region:
```
-- BLIS_Partition[i][0] is a subregion of BLIS_Partition[i] with the correct size
sizes = for i in [0, N): get_subregion_size(BLIS_Partition[i][0])
total_size = reduce_future_map(sizes, +)
R = create_region( [0, total_size) )
P = create_partition_by_weights(R, sizes)
for i in [0, N): translate(BLIS_Partition[i][0], P[i])
```
The code launches an index task to convert sizes of BLIS subregions into a future map. Then, it uses the future map to create an output region and a partition whose subregions respectively have the same sizes as their BLIS counterparts, and finally copies data from the BLIS partition to the output partition.

This conversion code is a perfectly valid Legion code, but the last step is somewhat ugly. We know that each pair of the subregions have the same size, but we can't use one's instance for the other only because their index spaces are different, and we should instead make a copy. What we really wanted here is a read-only view to the existing region and we could support such a read-only alias by extending `AttachLauncher`/`IndexAttachLauncher` as follows:

```
struct AttachLauncher {
  AttachLauncher(const RegionRequirement &target_requirement);
  void attach_region(const RegionRequirement &source_requirement);
};

struct IndexAttachLauncher {
  IndexAttachLauncher(const RegionRequirement &target_projection_requirement);
  void attach_partition(RegionRequirement &source_projection_requirement);
};

class Runtime {
  void create_read_only_alias(Context ctx, const AttachLauncher &launcher);
  void create_read_only_alias(Context ctx, const IndexAttachLauncher &launcher);
};
```
For `AttachLauncher`, the source and target regions must have the same volume, and for `IndexAttachLauncher`, each pair of the source and target subregions after projection must have the same volume. For both, the source and target requirements must have the same number of fields.

Here is how the attach operation works: When an attach operation `attach(R -> S)` is mapped, where `R` is the source and `S` is the target, the mapper picks one of the `R`'s instances, which will then be registered as a valid instance of `S`. The attach also invalidates all the existing instances of `S`. After mapping, the chosen instance of `R` can be used to map only read-only requirements and is invalidated when `S` is updated. For accessors, I believe the runtime has all the metadata to construct a mapping from the point of `S` to `R`'s automatically, even when `R` and `S` have different dimensions and/or have sparse index spaces. (having `R` and `S` have different dimensions effectively makes this a reshaping operation, which I think can be useful on its own.)

### Future-based approach

@lightsighter recently talked about allowing futures to hold Realm instances and one obvious use case is to make tasks return variable size outputs. Here is my version of how it should work.

First, we use `DeferredBuffer` to create Realm instances in tasks (Requirement 1). As far as I know, Realm instances are not directly exposed in the Legion API and `DeferredBuffer` is the closest thing that is available. `DeferredBuffer` also provides an accessor to the instance, which I believe cannot be created without a physical region at the moment.

Second, we introduce a new kind futures that keep Realm instances as values, which I will call *instance futures*. Today a Realm instance in a `DeferredBuffer` cannot outlive the task in which it was created, but we can make it escape by attaching it to an instance future that will hold a reference (Requirement 2). Instance futures are somewhat similar to deferred futures I described previously, but they can be used only with task launchers, which makes them much safer to use. Here are definitions of the instance future and the *instance map* (i.e., a map of instance futures):

```
class InstanceFuture : public Future {
  template <typename T, int DIM, typename COORD_T>
  void set_instance(const DeferredBuffer<T,DIM,COORD_T> &buffer);

  template <typename T, int DIM, typename COORD_T>
  DeferredBuffer<T,DIM,COORD_T> get_instance(void);
};

class InstanceMap : public FutureMap {
  InstanceFuture get_instance_future(const DomainPoint &point) const;
};  
```
Note that `InstanceFuture::get_instance` returns a deferred buffer so that users don't need to worry about creating accessors to Realm instances.

Just like a deferred future, the ready event of an instance future is triggered when the future is assigned an instance, and each instance future can be set only once. Mutating instance futures and instance maps requires privilege, for which we add the following methods to the task launchers:
```
enum InstancePrivilegeMode {
  INITIALIZE,
  READ_ONLY,
  READ_WRITE,
};

struct TaskLauncher {
  void add_instance_future(InstanceFuture future, InstancePrivilegeMode privilege);
};

class IndexTaskLauncher {
  void add_instance_map(InstanceMap map, InstancePrivilegeMode privilege);
};
```
A task must hold the `INITIALIZE` privilege to be able to mutate the instance future with `InstanceFuture::set_instance`. Passing an uninitialized instance future to a task with a privilege other than `INITIALIZE` is illegal. **EDIT**: Passing an initialized instance future with the `INITIALIZE` privilege is also illegal and every uninitialized instance future must be assigned an instance by the first task that takes it. Tasks with `READ_ONLY` or `READ_WRITE` privileges can make accesses to the instances in the futures that have been initialized. (`WRITE_ONLY` is  omitted here, since semantically it is not different from `READ_WRITE`. I'll get back to this in the last paragraph.) The futures in an instance map behave like point futures; i.e., they are individually assigned to the point tasks. Instance maps therefore must cover all the point tasks; i.e., the domain of the instance map must be a superset of the launch domain.

Alternatively, we could make tasks directly return instance futures, instead of having futures initially decoupled from instances. That will definitely make instance futures safer to use, but I haven't found a good way to handle cases where a task needs to return multiple instance futures.

A side benefit of introducing instance maps is that they can be used as an efficient abstraction for disjoint partitions of unordered data. The dependence analysis on futures is based on names, and thus can be performed much more efficiently than the region tree-based analysis (although it will still need to manage epoch lists).

Lastly, we will sometimes want to convert an instance future back to a region, or an instance map into a partition, in particular if we want to repartition the data (Requirement 3). This is similar to the "compaction" of the BLIS structure described in the previous approach, and my idea to address this is also similar:

```
struct AttachLauncher {
  AttachLauncher(const RegionRequirement &target_requirement);
  void attach_instance_future(std::map<FieldID, InstanceFuture> instances, bool own=false);
};

struct IndexAttachLauncher {
  IndexAttachLauncher(const RegionRequirement &target_projection_requirement);
  void attach_instance_map(std::map<FieldID, InstanceMap> instance_maps, bool own=false);
};

class Runtime {
  void attach_instances(Context ctx, const AttachLauncher &launcher);
  void attach_instances(Context ctx, const IndexAttachLauncher &launcher);
};
```

We extend the attach operation to take instance futures and convert them into region fields. Each instance must have the same volume as the target so that it can be "reshaped" into the target index space. The index attach operation works similarly, except that it requires instances in each instance map to have the same volumes as their corresponding target subregions. The `own` parameter determines whether the ownership of instances will be transferred to the target region; when `own=false`, the attach operation creates a read-only alias to the instances; when `own=true`, the target region becomes the owner of the instances and the instance futures lose the ownership and are invalidated. Passing an invalidated instance future to a subsequent task is illegal.

A major issue with the future-based approach is that it introduces an extra stream of allocations that is not in line with the mapping wavefront. As we discussed briefly in #841, a deferred buffer in a task `T` can cause a deadlock if the runtime assigns a chunk of memory that is committed to some future task that depends on `T`. A short term workaround that we discussed was to manage two pools of allocations, one for allocations from the mapper and another for those from tasks. We will need a better resource management scheme that doesn't require such a rigid separation, since the right boundary between the two pools isn't straightforward to choose and may even need to be adjusted dynamically.

Another thing to think about is what we should do if a task is mapped to a processor that has no access to the memory where the instance in a future locates. The instance future as designed in this writing holds a specific Realm instance, which makes it valid to use only in a certain set of tasks. In this design, the runtime would raise an error if the mapper mapped a task to a processor where some of its instance futures are invisible. Alternatively, we can build an abstraction around Realm instances so that they can be relocated. This will require that there be a mapping phase for instance futures, where the mapper either chooses one of the existing Realm instances or creates a new one. Under this scheme, `InstancePrivilegeMode::WRITE_ONLY` is semantically different from `InstancePrivilegeMode::READ_WRITE`, as the former doesn't require the valid data to be copied to a new instance.

---

I tried to enumerate all potential issues that I know of in each of the approaches, but please make comments if I missed any. Also, it would be great if any of you could find other approaches that are not listed here.

**Comment by lightsighter** [comment]
So in my mind, the fundamental issue that makes all of these approaches painful is this:

> A major issue with the future-based approach is that it introduces an extra stream of allocations that is not in line with the mapping wavefront. As we discussed briefly in #841, a deferred buffer in a task T can cause a deadlock if the runtime assigns a chunk of memory that is committed to some future task that depends on T.

Unbounded output and deferred execution do not play nicely together because they can lead to resource deadlocks like the one described above. I think the real question we need to be addressing is what we want the solution to the resource deadlock problem to be. Once we resolve that, then I think it will be much easier to figure out what programming model and API we want to encourage for unbounded output. For example, both the BLIS style scheme and the future-based approaches get much easier if you can break the mapping dependences that exists solely for avoiding resource deadlock.

On this front, I think the long-term solution is that we should optimistically execute assuming that resource deadlock never occurs. When it does we should detect it and then use the resilience mechanisms from #636 to roll-back and recover in more conservative way to ensure that resource deadlock does not occur again. Resilience is still a little ways off though, so I think we will need a different solution in the short term. Certainly the "two pools" approach is at the top of the list where you have a pool for deferred allocations and a separate pool for eager allocations. The problem with this approach is that the client needs to size these pools appropriately. It is safe to move memory from the eager pool to the deferred pool, but once it moves it cannot go back. If an application is bursty and needs to shuffle memory between the two pools dynamically we might look like we run out of memory when in reality the working set of the application actually can fit in memory. I'm open to other suggestions, but I think this needs to be the first point of the discussion before deciding what programming model and API we want to encourage.

All that being said, I think my plan for futures backed by instances looks quite different than the one @magnatelee has laid out here. I personally don't want to change the future API hardly at all. The idea is that Legion's existing futures would be backed by Realm instances internally. In practice, that likely just means that existing memory for futures (currently allocated by malloc) will be wrapped by a Realm external instance so that Realm can do copy operations to/from it. In some cases we may actually allocate instances (e.g. in true deferred cases) but that is an implementation detail. The important thing is that the Futures as they exist today will be forwards compatible with anything that comes later. There should be no need for a new future type and I'll be pretty resistant to introducing new data types that are not our existing futures or regions.

I too also thought about using attach operations to convert futures into regions, effectively reinterpreting the unstructured future data as a structured region type. I'm pretty ambivalent about what the API would look like for this, but the one thing I'm adamant about is that this reinterpret operation would need to "consume" the futures, i.e. the futures would no longer be valid after the reinterpret operation into regions. There is currently no way for Legion data types to "alias" right now and I want to maintain that invariant both for simplicity of the programming model and simplicity of the implementation. I don't see any value in having them be aliased, requiring the user to manage that relationship, and/or requiring the runtime to maintain coherence across the relationship. 

One side note is that `InstanceMap` object that @magnatelee proposed is pretty similar to the `PhysicalRegionMap` object proposed for #550 which was something I was looking for secondary use case for.

**Comment by lightsighter** [comment]
Here is a useful mental framework to use when thinking about the tension between deferred execution and unbounded memory outputs. There are three objectives we want to attain:

1. We want deferred execution. We should never have to block to introspect the results of a deferred computation to make forward progress.
2. We want memory to look "full size". It's still possible to run out of memory just like it is in the case where we do eager execution, but we want to guarantee that the deferred execution of the program will run out of memory if and only if the in-order eager execution of the program also runs out of memory.
3. We want to not hang due to resource deadlocks.

In most cases we can only get 2 out of 3. The optimistic eager deferred execution with resilience recovery gets all three. The two pools approach gets objectives 1 and 3, but not 2. Objective 3 is pretty much required because hangs are bad for users. Objective 1 is strongly desired for performance.

**Comment by manopapad** [comment]
I think there's also objective 4: avoid duplicate computation, that the resilience recovery scheme does not fulfill (we potentially have to restart the task). Actually the two-pass scheme poses a similar tradeoff (achieves 1,2 and 3 but not 4).

Adding some solutions I was discussing with Mike. Neither is great, as they both have the potential for duplicate execution, but similar to the resilience recovery scheme they are efficient in the happy path:

#### Optimistic two-pass

This solution assumes an eager pool that's a good bit smaller than the deferred pool. The unbounded-output task `T` tries to fulfill its memory requirements from the eager pool. If the allocation succeeds then the task completes successfully and returns a pointer to its output memory. If the allocation fails then the task returns the size it actually needed, and a copy `T'` that was predicated on `T` failing asks for that amount of memory from the deferred pool (i.e. we fall back to the two-pass scheme).

This approach still involves a separate eager pool, so not all memory is utilized, and  would require some changes so we can return two futures from a task.

#### Doubling approach

This one is less practical, but adding it here for completeness.

Launch a sequence of identical copies of the unbounded-output task `T`, each predicated on the return value of the previous one. The first copy is assigned a pre-sized index space of say 1M elements, served normally from the deferred pool. If that proves to be enough then we return `false`, inform the runtime of how much memory we didn't use, and the dependent tasks do not run. Otherwise return `true` and the next task, which was assigned an index space of 2M elements, runs etc.

This solution would cause the least amount of blocking because all index space sizes are known when the top-level task executes, but it might take multiple tries to succeed. Also, by the time we can reclaim `T`'s unused memory the subsequent tasks will have been mapped and scheduled assuming `T` used its full allocation, thus constraining task parallelism.

**Comment by magnatelee** [comment]
Another approach I can think of is the optimistic mapping approach. If an eager allocation from a task `T` is failed, the runtime cancels mapped tasks in reverse chronological order until it finds a big enough chunk to give out. After fulfilling `T`'s request, the runtime tries to remap the canceled tasks, this time with `T`'s completion taken into account. Since the runtime maps ahead only one scheduling window worth of tasks, there is an upper bound of the amount of cancellation and also the amount of bookkeeping the runtime needs to do.

Comparing with the optimistic task execution, it's the cost of remapping canceled tasks vs. the duplicate computation due to repeated execution, and I'm expecting the former to be smaller than the later in realistic cases.

**Comment by manopapad** [comment]
Optimistic mapping sounds like the best approach so far.

**Comment by lightsighter** [comment]
There are two problems with the optimistic mapping approach. The first problem is that this statement is not true:

> Since the runtime maps ahead only one scheduling window worth of tasks, there is an upper bound of the amount of cancellation and also the amount of bookkeeping the runtime needs to do.

Tasks can produce sub-tasks and those sub-tasks can produce sub-tasks and so forth, so when you're rolling back just one task in your context, you are actually rolling back an unbounded number of tasks in the sub-tree of tasks spawned by that one task. If you constrain your claim to say that all tasks are using leaf-variants then it becomes true, but if any of those tasks are non-leaf tasks then the task you need to roll back so it can release it's resource could have started running and really what you need to preempt is the mapped sub-tasks produced by that parent task. Either the number of those mapped sub-tasks (or mapped sub-tasks in those sub-tasks) you might need to preempt are unbounded or you're talking about rolling back the entirety of a partially executed task which puts you right back in the world of #636. Pick your poison.

The second problem with the optimistic mapping approach is that what @alexaiken pointed out in the meeting: recovery needs to be fast in this case since we're likely going to be doing it often in memory constrained environments. It's unclear how much work and effort it's going to be to make this fast enough that it is off the critical path in most cases.

I'm not saying we can't do this approach, but I want to make sure we are weighing all the pros and cons and not looking at some solutions with rose colored glasses.

**Comment by magnatelee** [comment]
I thought a bit more about the optimistic mapping approach and I think we don't even need to cancel and remap tasks when an eager allocation fails. What we need here is a mechanism to "borrow" instances from mapped tasks. The borrowing operation amounts to 1) finding a set of instances to borrow for a given failed allocation and 2) adding the borrower's completion event to the set of preconditions of each borrowed instance. Step (1) will need some guidance from Realm as we ideally want to borrow only a minimal set of contiguous instances. Step (2) will be a much simpler operation than both cancel-and-remap for the optimistic mapping and rollback-and-retry for the optimistic deferred execution. Also, the borrowing can't cause any deadlock because the lender either has no dependence on the borrower (in which case the added dependence is spurious) or is already a successor of the borrower in a dependence chain (in which case the added dependence is simply transitive). 

I'm sure the borrowing approach will be more difficult to implement than it sounds, but at least it sounds much more approachable and feasible than the two optimistic approaches.

**Comment by lightsighter** [comment]
Be very careful, this statement is not true if you have a task tree with more than two levels:

> Also, the borrowing can't cause any deadlock because the lender either has no dependence on the borrower (in which case the added dependence is spurious) or is already a successor of the borrower in a dependence chain (in which case the added dependence is simply transitive).

For step (1), how are you going to handle the race between doing the borrow analysis and the task starting running because it has no dependences on the task that needs memory?

I think you might want to investigate how Realm's deferred allocation logic works before going to far down this path. There are more cases there then you might think already and getting them to operate correctly with "borrowing" is going to be quite complex. I think it could probably be done, but I would not want to do it without a really rigorous test suite to try to exercise all the various different cases.

**Comment by magnatelee** [comment]
> Be very careful, this statement is not true if you have a task tree with more than two levels:

I think it's still true even for a task tree with multiple levels; unless one executing task is trying to borrow from some other executing task (the case that I'll discuss next), there is no way you can create a cycle, as the lender will never have a chance to get started before the borrower terminates. The real danger I think is the case where the borrowed instance is escaping the borrower's context; if an instance `I`  borrowed from task `T` is passed to another task `U` that is a dependent successor of `T`, it will cause deadlock. I'll need to think more how to handle such cases.

Coming back to the case where one executing task needs to borrow from another executing task, I think in that case the borrower needs to get blocked until one of the executing tasks retires. Here is why: the scheme I'm thinking of will start to borrow from mapped-but-not-executing tasks first. If we still need to borrow from one of the executing tasks even after borrowing from all the mapped-but-not-executing tasks, it means that size of the working set is too big to have the borrower run concurrently, or that the working set simply doesn't fit to memory. In either case, the right choice is to have the borrower blocked on the allocation and retry. We should also forbid borrowing from the blocked task, as doing so will lead to a deadlock. 

> For step (1), how are you going to handle the race between doing the borrow analysis and the task starting running because it has no dependences on the task that needs memory?

The borrowing analysis and the scheduler can atomically change the task state like this: if the borrowing analysis gets to a yet-to-execute task first, it  changes the task's precondition and delays the execution. If the borrowing analysis is late and the runtime has already scheduled the task, the borrowing analysis moves on to other tasks and falls back to the borrowing-from-an-executing-task case that I described above.

> I think you might want to investigate how Realm's deferred allocation logic works before going to far down this path. There are more cases there then you might think already and getting them to operate correctly with "borrowing" is going to be quite complex. I think it could probably be done, but I would not want to do it without a really rigorous test suite to try to exercise all the various different cases.

Thanks for a suggestion, although I wouldn't do this with a rigorous test suite, but rather with a rigorous formalism. :)

**Comment by lightsighter** [comment]
> The real danger I think is the case where the borrowed instance is escaping the borrower's context; if an instance I borrowed from task T is passed to another task U that is a dependent successor of T, it will cause deadlock. I'll need to think more how to handle such cases.

This is exactly what I mean when I say that your approach doesn't work if the task tree has more than two levels. Once you have more than two levels then instances can flow back up the tree (and then maybe down again) which is when you get into trouble.

> Here is why: the scheme I'm thinking of will start to borrow from mapped-but-not-executing tasks first.
> The borrowing analysis and the scheduler can atomically change the task state like this: if the borrowing analysis gets to a yet-to-execute task first, it changes the task's precondition and delays the execution.

Something else to be careful of here: what you are proposing is an approach that spans the Legion/Realm interface. Realm has information about what instances are allocated and when they are available with what events. Legion has information about what tasks are mapped and the regions associated with them. Which side of the interface are you thinking to implement this on?

> Thanks for a suggestion, although I wouldn't do this with a rigorous test suite, but rather with a rigorous formalism. :)

My point is, regardless of whether you have a formalism or not, you still have to write the code which involves a lot of very fine grained locking to be fast, and that fine grained locking can be fraught with race conditions and deadlocks, independent of the correctness of any higher-level protocol it is trying to implement.

**Comment by elliottslaughter** [comment]
A couple thoughts, in no particular order:

It seems to me that borrowing can only be safe if the instance does not contain valid data. Otherwise Legion would have to confirm that all (valid) data in the region is contained in at least one other instance, which I believe is a difficult problem. Or we'd have to write the data to disk, but that can be potentially slow. I haven't seen this discussed in the thread so I just want to make sure people are thinking about this aspect of the problem.

In terms of the interfaces proposed, none of the suggestions here strike me as particularly Legion-like. I'm not quite sure what I'd do instead but I resonate with Mike's desire not to change the Future interface, and for conversion operations (if any) to be one-time, consuming operations that transform the data into normal regions/partitions/etc. I definitely want to avoid adding another interface for passing instances to tasks, outside of regions. That would split the entire API between regions and these new like-regions-but-not-the-same data structures we're introducing.

Can we make a constraint where variability can be handled within a single task (i.e. a leaf)? If so that seems to simplify the design space somewhat. If not then we should probably talk about the exact requirements we need to support.

**Comment by magnatelee** [comment]
> Something else to be careful of here: what you are proposing is an approach that spans the Legion/Realm interface. Realm has information about what instances are allocated and when they are available with what events. Legion has information about what tasks are mapped and the regions associated with them. Which side of the interface are you thinking to implement this on?

I think it should be implemented on Legion, since we need to reason about task dependencies and copies to make instances valid, but if we need to borrow more than one instance for a give request, we will need to query instances' physical locations to Realm to find a contiguous chunk that is big enough from a set of instances eligible to borrowing. When Legion updates the readiness of an instance due to borrowing, I think it needs to tell Realm the new ready event so that Legion and Realm are on the same page.

> My point is, regardless of whether you have a formalism or not, you still have to write the code which involves a lot of very fine grained locking to be fast, and that fine grained locking can be fraught with race conditions and deadlocks, independent of the correctness of any higher-level protocol it is trying to implement.

Your point is well taken and I didn't underestimate the potential complexity of an efficient implementation. What I was trying to say is that the borrowing approach, or any approach that mixes immediate and deferred allocations, seems to be in the territory that we have never been to, and so even a high-level protocol would need a rigorous reasoning and formalism, which would help us better grasp the problem.

> It seems to me that borrowing can only be safe if the instance does not contain valid data. Otherwise Legion would have to confirm that all (valid) data in the region is contained in at least one other instance, which I believe is a difficult problem. Or we'd have to write the data to disk, but that can be potentially slow. I haven't seen this discussed in the thread so I just want to make sure people are thinking about this aspect of the problem.

I agree with your reasoning. We should borrow only the instances that are yet to be created, i.e., whose names are created by Realm under some ready events but that are not yet ready. For this, the borrowing analysis needs to reason about "liveness" of instances in the future point in execution as well so that it doesn't borrow live instances (which are in fact a part of the working set). In case where the borrowed instance was made valid by some Ream copies, we will need to update the preconditions of those copies and not the instance's ready event, which will practically amounts to remapping of the corresponding region requirement and be something we'd like to avoid unless necessary.

> In terms of the interfaces proposed, none of the suggestions here strike me as particularly Legion-like. I'm not quite sure what I'd do instead but I resonate with Mike's desire not to change the Future interface, and for conversion operations (if any) to be one-time, consuming operations that transform the data into normal regions/partitions/etc. I definitely want to avoid adding another interface for passing instances to tasks, outside of regions. That would split the entire API between regions and these new like-regions-but-not-the-same data structures we're introducing.

Despite the lengthy text I wrote, I actually feel the same way with you guys. That said, one thing I'd like us to think about is what the interface should be to express tasks returning multiple instances. So far, we've been avoiding having this interface simply by doing the following program conversion, which works well only when we know the size of region:

```
a, b= f(); ==> a = region(size_a); b = region(size_b); f(a, b)
```

If this is the pattern we will encourage for variable size task outputs (and we don't use the BLIS approach), we will need to allow unbounded regions to be created and have their sizes later determined by the first callee task. Inside the task we will need to provide some form of allocator. And if this is an index task, we also need to create partitions on those unbounded regions and after the point tasks are done, we need to recover the global numbering of the regions and subregions. All this is actually the same thing that I described in [this comment](https://github.com/StanfordLegion/legion/issues/707#issuecomment-584016892), and my impression is that we don't want to go this route.

If we do the same conversion above but with something other than regions, say futures, then that becomes the interface that I proposed in this issue.

If we are going to have tasks return multiple values in futures, we will at least need to provide some combinators on futures so that the users can use them easily. Futures are monadic values, and without providing combinators peaking inside the monad, users will have to write a ton of boilerplate tasks. For example, the return type of a task returning two values is practically `Future<pair<T1,T2>>` (or `FutureMap<pair<T1,T2>>` if this is an index task), and the user needs to launch a task to convert the pair into each of the values. The new local task support can make this efficient, but can't save effort to write the task (though Regent can automate this a lot). I feel we should at least provide task preambles and postambles so that if the user wrote a task returning a value of `tuple<T1, ..., Tn>`, he could see a value of `tuple<Future<T1>, ..., Future<Tn>>` from the task launch.

> Can we make a constraint where variability can be handled within a single task (i.e. a leaf)? If so that seems to simplify the design space somewhat. If not then we should probably talk about the exact requirements we need to support.

I was actually going to propose something similar; we could add a mapper interface so that in the mapper the user can specify a lower bound of the amount of eager allocations in each task. Part of the reason that eager allocations are hard is that the tasks didn't tell all about their memory usage. So the idea is to have the mapper specify some scratch pad memory space that can be used for eager allocations. The memory space unused after task execution is reclaimed by Realm, and when the task runs out its scratch pad space, it falls back to the borrowing. The borrowing approach also becomes easier with scratch pad allocations, since they take much less to borrow than normal instances, which may have Realm copies attached to them.

This may sound similar to the two-pool approach, but is actually different in two ways: First, all scratch pad allocations are coming from the same pool as normal instance allocations. Second, we only ask the user to specify lower bounds, not upper bounds. (And the lower bound being 0 is a perfectly valid mapper output; it only makes things more expensive.)

Alternatively, we could add a mode where the mapper specify upper bounds, which will be still better than the vanilla two-pool approach, since all the allocations are still in line with the deferred allocation scheme and there will be no eager allocations inside tasks.

Even if we don't do this scratch pad idea, I think requiring the mapper to specify whether the task is ever going to make any eager allocations will be helpful for anything we do in the runtime, since we don't want to pay for extra bookkeeping in cases where there is no task with variable size outputs.

**Comment by lightsighter** [comment]
```
a, b= f(); ==> a = region(size_a); b = region(size_b); f(a, b)
```

I would argue that this pattern is supported today, it's just that you need to take the time to pack your values into a struct which is returned in the future. Packing and unpacking values to/from structs is annoying, but not a fundamental pattern that Legion is prohibiting. The place where this gets into trouble is if `a` and `b` are dynamically sized collections and not just statically sized data types. Even still we have some support for collections today (e.g. the `legion_serialize*` method on future types), but it comes at the cost of additional packing/unpacking overhead. So the only reason to have multiple futures being returned from a task is if you want to support returning multiple collections of data from a task without any overhead. It sounds like that is the problem we have here, but I want to be clear that that is the reason that we are discussing this and not just that we can't do this at all today. 

Now in general we encourage users to put collections of data in regions. Rather than trying to graft all this onto futures and then convert them to regions, I'm wondering if it isn't just better to have Legion support some "output" region kinds that are "append only" or something like it that will construct the index space and an initial partition based on the output regions. Notice though that this comes back to the decoupling of API design from resource deadlock issues I brought up before. This approach though still suffers from resource deadlock if we don't resolve that, so I would rather focus on how want to address resource allocation first before we figure out what we want the API to be.

> If we are going to have tasks return multiple values in futures, we will at least need to provide some combinators on futures so that the users can use them easily.

I'm strongly against this approach. It is much easier to supply multiple return futures from a task than to guess and provide all the kinds of combinators people might ultimately want. Again though, let's ignore API design until we figure out how we want to allocate resources. Backing futures with instances means that both regions and futures are using the same Realm primitive for resource allocation so we can think entirely about how to manage instance allocation with different timelines of allocations/deletions.

> I was actually going to propose something similar; we could add a mapper interface so that in the mapper the user can specify a lower bound of the amount of eager allocations in each task.

This is nearly trivial to add and is conceptually simple to reason about. I think we can safely ignore this for now and assume the lower bound is set to `0` so that we can figure out what we're going to do in the hard case when we overflow this lower bound. We'll be able to add in the lower bound case whenever we implement this, but it still leaves open the question of both correctness and performance when you do have to go somewhere else to find memory to use.

One last thing: let me just complicate this even further by pointing out another interaction that this issue has. For #828 the runtime will need to be allocating shadow instances on the fly for handling variable sized field data. Ideally, this will happen in a new separate stage of the pipeline to avoid blocking the mapping process while we compute and fill in offsets into the variable sized field shadow instances. This will create a third timeline of allocations/deletions. Three is all I see for now, but I won't rule out that other stages might ultimately need their own memory allocations as well in the future, so now we're really talking about `N` timelines of allocations/deletions that we might need to reason about.

**Comment by magnatelee** [comment]
> Rather than trying to graft all this onto futures and then convert them to regions, I'm wondering if it isn't just better to have Legion support some "output" region kinds that are "append only" or something like it that will construct the index space and an initial partition based on the output regions.

This sounds pretty much like what I proposed in my comment to #707, though the exact wording is slightly different, and I'm in favor of this. 

> This is nearly trivial to add and is conceptually simple to reason about. I think we can safely ignore this for now and assume the lower bound is set to 0 so that we can figure out what we're going to do in the hard case when we overflow this lower bound. 

Like I said in our private conversation, I don't see why we can't start implementing the upper bound scheme today, if it's nearly trivial to add. As much as I like to have a good long term solution for multiple allocations streams in the deferred execution setting, I want a near-term solution, even if it's partial, so that we can have people, including myself, make progress less painfully than they do now. And I believe the upper bound scheme is part of what we will need to implement for anything that will come later, so it won't be entirely a waste effort. This will also make our interface discussion more relevant and real, as the users don't care how their in-task allocations are actually catered, but they will definitely want a reasonable interface.

**Comment by elliottslaughter** [comment]
I guess I'm a little confused. How is the upper bound scheme any different from what we support today? Isn't it just sugar on regions, virtual mapping, and partitioning/inline mapping inside a child task?

Why couldn't this be a library?

I'm ok with adding new features to Legion when they're actually necessary, but if this can be done as a library I think we should build the library first, and then decide what (if anything) to move into the runtime later.

**Comment by magnatelee** [comment]
We can't do it as a mere syntactic sugar around the BLIS approach, which I've already done and I believe is not the right solution for the reasons I listed above. What I think we need is a combination of 1) a mapper interface to take the upper bound of scratch pad space at the mapping phase, 2) a mechanism for intra-task allocation in that scratch pad space, and 3) a way to assign a region name to the allocated instance. And a bonus point would be 4) a Realm mechanism to reclaim over-claimed scratch pad space. And I believe all these mechanisms (maybe except 4) can be done reasonably soon and will still be used even after we switch to the lower bound scheme.

**Comment by elliottslaughter** [comment]
I guess that approach is fine but it still seems like we're adding a lot of moving parts, and they're not entirely orthogonal.

What do we actually need beyond user-level BLIS? I would say:

1. Legion mechanisms for deferring certain region creation operations without needing to block the issue of dependent operations
2. Realm mechanisms to growable memory consumption inside a task (what you call a scratchpad)
3. Realm mechanisms for resetting the index space of an index space, so that you can use it as part of a region later on without copying or repartitioning

In particular, it seems to be that (2) can be entirely implemented at user level without help of either runtime. Also, (3) is strictly a performance optimization and with partitioning/compaction you could avoid needing this (at some cost to performance). It seems (1) is the only real fundamental advance here, and how we deal with deferred, dynamic memory allocation.

I just offer this as a framework because it seems like we're entangling multiple things into this proposal, and again I want to minimize the functionality we actually need to add to either runtime to support this. Also, if we can identify the orthogonal pieces, maybe there are better ways to implement some of them.

For example, an alternate way to implement (3) would be to add a "reshape" operator to Realm. Would that be better than adding a whole suite of scratchpad operations? (I don't know, but it seems like we should at least think about it.)

**Comment by magnatelee** [comment]
In my opinion, (2) in your list is what this issue is all about, allowing dynamic allocation inside a task and having it returned from the task, and I don't think it can be implemented completely at user level, unless we impose a hard limit on the growth at the mapping phase. The only reason that inline mappings of subregions looked safe with BLIS is that it blocked dependent operations, and as soon as we have them unblocked, those inline mappings all fall back to yet another form of eager allocations.

I think the root cause that makes variable size data difficult is that the default mode of the runtime analysis requires the shape of data in the form of regions, even for cases where the runtime merely passes a set of instances from one task to another. So, I look at item (1) in your list as a new kind of region/future/etc. that the runtime treats as an opaque data object, and in this regard, Mike's initial proposal of instance-backed futures makes sense.

> For example, an alternate way to implement (3) would be to add a "reshape" operator to Realm. Would that be better than adding a whole suite of scratchpad operations? (I don't know, but it seems like we should at least think about it.)

I want to point out that reshaping is actually orthogonal to being able to dynamically allocate data within a task and have it escape the task, which is what scratchpad is really about.

**Comment by manopapad** [comment]
I got a little lost in the discussion (I was oblivious to the intricacies of rolling back task scheduling), but I think as a temporary measure the upper-bound scheme (each task declares an upper bound on the amount of memory it will allocate, and whatever is not used is collected at task completion) would be sufficient for my use cases. Especially if it could play nicely with predicated execution so you can recover from underestimating your memory needs without crashing the application (i.e. what I was alluding to in the "doubling approach").

**Comment by elliottslaughter** [comment]
> In my opinion, (2) in your list is what this issue is all about, \[...] and I don't think it can be implemented completely at user level, unless we impose a hard limit on the growth at the mapping phase.

Sorry, I was responding to your comment above about implementing an upper bound approach first. I agree that once you no longer have an upper bound this cannot be done (safely) at user level.

But I guess I still don't see why there has to be a scratchpad/instance distinction. Legion has to do something to manage the potential resource deadlock. But from Realm's perspective these seem like two very non-orthogonal interfaces (isn't what you call a scratchpad just a superset of the current instance interface?). I'm just trying to improve the orthogonality of the interfaces here.

**Comment by streichler** [comment]
Chiming in with some Realm-related considerations.

First, two things that almost certainly will not be added:
1. A generic "realloc" that can grow an instance.  In practice, any growing realloc is implemented by malloc'ing a new buffer, copying the contents, and the deleting the old one, so there's no performance advantage to having that be one Realm operation instead of three.
2. A "reshape" operation that changes the layout of an existing instance.  The way that instance layout metadata is shared across nodes makes this really hard.

The thing I could reasonably implement is a "shrink" operation that has the following semantics:
1. Destroy an existing instance, with whatever layout it had, based on whatever precondition you give.
2. Atomically create a new instance that uses some contiguous subset of the storage previously dedicated to the deleted instance, using whatever layout.  The contents of the reused subset are guaranteed to be preserved and it's guaranteed that no copy will be performed.  The creation would have a completion event that may or may not be the same as the precondition.
3. Whatever storage that was dedicated to the deleted instance that is not used by the new instance is made available for future instance allocations.

**Comment by magnatelee** [comment]
> But I guess I still don't see why there has to be a scratchpad/instance distinction. Legion has to do something to manage the potential resource deadlock. But from Realm's perspective these seem like two very non-orthogonal interfaces (isn't what you call a scratchpad just a superset of the current instance interface?). I'm just trying to improve the orthogonality of the interfaces here.

Sorry for not being clear about what I meant by the scratchpad. From Legion's perspective, a scratchpad is just a Realm instance with an additional interface that allows it to release the unused space. So, you're right that we don't need and want strict distinction in the interface, but we will just need to add a couple of additional functionalities to the current interface.

> A "reshape" operation that changes the layout of an existing instance. The way that instance layout metadata is shared across nodes makes this really hard.

I believe the reshaping can be done purely on the Legion side by establishing a mapping between the old index space and a new one (though for sparse regions the mapping can require a O(N) description in the worst case) and teach the accessor how to use that mapping. In case of constant amount translation of a dense region, which I'm expecting will be the majority of the use cases, this mapping will be trivial.

> The thing I could reasonably implement is a "shrink" operation that has the following semantics:

That sounds exactly what we would need for the scratchpad, though Legion will need to migrate users on the old instance name to the new one.

**Comment by lightsighter** [comment]
> I think the root cause that makes variable size data difficult is that the default mode of the runtime analysis requires the shape of data in the form of regions

That's not what makes it hard. It's hard because if you don't do it right you get a resource deadlock. While any interface we choose will require some non-trivial effort to implement, I don't think the distinction about how we end up capturing (or not) the structure of the data will cause any significant difficulties. The hard part is allocating space for that data without blocking deferred execution. I continue to believe that is the only thing we should be focusing on right now.

> But I guess I still don't see why there has to be a scratchpad/instance distinction.

Echoing what @magnatelee said above, to Realm these are all instances. A scratchpad interface in Legion would just be a way to have a heap of memory backed by a Realm instance for doing dynamic allocations that only live during the lifetime of the task. I'm ultimately moving Legion in a direction where literally every single memory allocation, both from the application, as well as the Legion runtime, will be backed by Realm instances, so having an interface for handling that was going to happen sooner or later anyway. This also maintains the framework that I set up before when thinking about the resource deadlock problem: think about everything as Realm instances. Whether those instances represent regions, futures, or some "scratchpad" memory is irrelevant.

> I believe the reshaping can be done purely on the Legion side by establishing a mapping between the old index space and a new one

It's only going to be "free" if you can construct an affine mapping between them.

> though for sparse regions the mapping can require a O(N) description in the worst case

It's not just sparse arrays:

```
x = np.zeroes((4,6))
y = x.reshape((3,8))
```

There's no affine mapping between the points in those two dense arrays, yet they have the same number of points and as far as NumPy is concerned would be backed by the same data.

> I think as a temporary measure the upper-bound scheme (each task declares an upper bound on the amount of memory it will allocate, and whatever is not used is collected at task completion) would be sufficient for my use cases

That's good to know ;) @magnatelee's use case is much harder. When you're doing a join if you have one table of size `M` and another table of size `N` then you can potentially have an output table that is `O(MN)`. In practice, both `M` and `N` are very large but the output is hardly ever `MN` which is huge, so you can't really afford to over-approximate too much.

**Comment by streichler** [comment]
>When you're doing a join if you have one table of size M and another table of size N then you can potentially have an output table that is O(MN). In practice, both M and N are very large but the output is hardly ever MN which is huge, so you can't really afford to over-approximate too much.

What technique is being used for the join?  A hash join with an interesting number of buckets ought to be able to tighten that bound quite a bit (i.e. by an expected factor of the number of buckets) without needing an extra pass over the data.

**Comment by magnatelee** [comment]
> That's not what makes it hard. It's hard because if you don't do it right you get a resource deadlock. While any interface we choose will require some non-trivial effort to implement, I don't think the distinction about how we end up capturing (or not) the structure of the data will cause any significant difficulties. The hard part is allocating space for that data without blocking deferred execution. I continue to believe that is the only thing we should be focusing on right now.

To clarify what I said, I was talking about where that hard part originated from. If the runtime treated everything as an opaque data object and tracked dependencies purely based on their identities, like in many other tasking runtimes, we didn't have to reason about multiple streams of allocations in the first place, as there will be one stream along the execution cadence. (Needless to say, those systems provide much weaker guarantees/automation for data coherence as a consequence.) So, while I agree with your comment about the difficulty and that we should all focus on it, I don't think that's source of the problem, and I think it is more on our decision to reason about the shape data prior to task execution.

> It's only going to be "free" if you can construct an affine mapping between them.

That's true. By the majority, I meant the cases with variable size outputs, which will almost likely use 1D index spaces.

> What technique is being used for the join? A hash join with an interesting number of buckets ought to be able to tighten that bound quite a bit (i.e. by an expected factor of the number of buckets) without needing an extra pass over the data.

(I'm using hash join.) That's true, but I think the assumption here is that the hash tables are available when computing the bound, which isn't my case, as I want to eventually ditch the two-pass scheme for joins (unless I find a sane way to share hash tables between the two passes).

**Comment by lightsighter** [comment]
> If the runtime treated everything as an opaque data object and tracked dependencies purely based on their identities, like in many other tasking runtimes, we didn't have to reason about multiple streams of allocations in the first place, as there will be one stream along the execution cadence.

This isn't true. In Legate Dask, all the data is stored in futures so it is all opaque blobs of bits with no structure, but we have the same problem. Futures are allocated on two different timelines: some futures are made eagerly as tasks finish and produce outputs while other futures are being allocated in a deferred mode as tasks are mapped onto remote nodes. You get two independent streams of allocation requests: one eager and one deferred. It doesn't have anything to do with whether we are tracking the structure of the program data or not; it only matters whether you are doing deferred allocation or not.

**Comment by streichler** [comment]
@magnatelee is this done?  If not, can you summarize what remains?

**Comment by magnatelee** [comment]
There is one extension I'd like to make, but the basic functionality is done and I'd feel comfortable closing this issue.

**Comment by magnatelee** [comment]
Actually, we are still waiting for the changes to be merged, so I guess we should keep this open until that merge happens.

**Comment by lightsighter** [comment]
Adding dependence on #765

**Comment by magnatelee** [comment]
> Actually, we are still waiting for the changes to be merged, so I guess we should keep this open until that merge happens.

I believe whatever the change was, it's long been merged, so closing this issue.
